crates/ruff/src/rules/pycodestyle/rules/logical_lines/extraneous_whitespace.rs

```diff
@@ -92,61 +102,137 @@ impl Violation for WhitespaceBeforeCloseBracket {
-            }
-            TokenKind::Rbrace
-            | TokenKind::Rpar
-            | TokenKind::Rsqb
-            | TokenKind::Comma
-            | TokenKind::Semi
-            | TokenKind::Colon => {
-                if !matches!(last_token, TokenKind::Comma | TokenKind::EndOfFile) {
-                    if let (Whitespace::Single | Whitespace::Many | Whitespace::Tab, offset) =
-                        line.leading_whitespace(token)
-                    {
-                        let diagnostic_kind = if matches!(
-                            kind,
-                            TokenKind::Comma | TokenKind::Semi | TokenKind::Colon
-                        ) {
-                            DiagnosticKind::from(WhitespaceBeforePunctuation)
-                        } else {
-                            DiagnosticKind::from(WhitespaceBeforeCloseBracket)
-                        };
-
+        if let Some(symbol) = BracketOrPunctuation::from_kind(kind) {
+            match symbol {
+                BracketOrPunctuation::OpenBracket(symbol) => {
+                    let (trailing, trailing_len) = line.trailing_whitespace(token);
+                    if !matches!(trailing, Whitespace::None) {
                         context.push(
-                            diagnostic_kind,
-                            TextRange::at(token.start() - offset, offset),
+                            WhitespaceAfterOpenBracket { symbol },
+                            TextRange::at(token.end(), trailing_len),
                         );
                     }
                 }
+                BracketOrPunctuation::CloseBracket(symbol) => {
+                    if !matches!(prev_token, TokenKind::Comma | TokenKind::EndOfFile) {
+                        if let (Whitespace::Single | Whitespace::Many | Whitespace::Tab, offset) =
+                            line.leading_whitespace(token)
+                        {
+                            context.push(
+                                DiagnosticKind::from(WhitespaceBeforeCloseBracket { symbol }),
+                                TextRange::at(token.start() - offset, offset),
+                            );
+                        }
+                    }
+                }
+                BracketOrPunctuation::Punctuation(symbol) => {
+                    if !matches!(prev_token, TokenKind::Comma | TokenKind::EndOfFile) {
+                        if let (Whitespace::Single | Whitespace::Many | Whitespace::Tab, offset) =
+                            line.leading_whitespace(token)
+                        {
+                            context.push(
+                                DiagnosticKind::from(WhitespaceBeforePunctuation { symbol }),
+                                TextRange::at(token.start() - offset, offset),
+                            );
+                        }
+                    }
+                }
             }
+        }
+
+        prev_token = kind;
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
```

Comment:
All this modeling down here is kind of verbose, but does capture the logic and semantics accurately.

Reply:
An alternative that removes the need for at least the `OpenBracket`, `CloseBracket` and `Punctuation` enums is to store a `char` in the `BracketOrPunctuation` variants. 

```suggestion
    OpenBracket(char),
    CloseBracket(char)
    Punctuation(char)
```

It's slightly less explicit but I think it would do the job, especially since this is all internal.

Reply:
Why are you trying to make invalid states representable???

Reply:
Because I'm lazy and it's private ;)